athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/AbstractStorageMetadata.java

```diff
@@ -0,0 +1,713 @@
+     */
+    protected AbstractStorageMetadata(StorageDatasourceConfig config) throws IOException
+    {
+        this.datasourceConfig = requireNonNull(config, "StorageDatastoreConfig is null");
+        requireNonNull(config.credentialsJson(), "GCS credential JSON is null");
+        requireNonNull(config.properties(), "Environment variables were null");
+        this.extension = requireNonNull(config.extension(), "File extension is null");
+        GoogleCredentials credentials
+                = GoogleCredentials.fromStream(new ByteArrayInputStream(config.credentialsJson().getBytes(StandardCharsets.UTF_8)))
+                .createScoped(Lists.newArrayList("https://www.googleapis.com/auth/cloud-platform"));
+        storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean containsInvalidExtension(String objectName)
+    {
+        return containsExtension(objectName)
+                && !objectName.endsWith(datasourceConfig.extension());
+    }
+
+    /**
+     * Return a list of Field instances with field name and field type (Arrow type)
+     *
+     * @param bucketName  Name of the bucket
+     * @param objectNames Name of the file in the specified bucket
+     * @return List of field instances
+     */
+    @Override
+    public List<Field> getTableFields(String bucketName, List<String> objectNames)
+    {
+        LOGGER.info("Retrieving field schema for file(s) {}, under the bucket {}", objectNames, bucketName);
+        requireNonNull(objectNames, "List of tables in bucket " + bucketName + " was null");
+        if (objectNames.isEmpty()) {
+            throw new UncheckedStorageDatasourceException("List of tables in bucket " + bucketName + " was empty");
+        }
+        LOGGER.debug("Inferring field schema based on file {}", objectNames.get(0));
+        String uri = createUri(bucketName, objectNames.get(0));
+        BufferAllocator allocator = new RootAllocator();
+        DatasetFactory factory = new FileSystemDatasetFactory(allocator,
+                NativeMemoryPool.getDefault(), getFileFormat(), uri);
+        // inspect schema
+        return factory.inspect().getFields();
+    }
+    /**
+     * Returns a list of all buckets from a cloud storage as databases
+     *
+     * @return List of database names
+     */
+    @Override
+    public List<String> getAllDatabases()
+    {
+        Page<Bucket> buckets = storage.list();
+        for (Bucket bucket : buckets.iterateAll()) {
+            String bucketName = bucket.getName();
+            String validName = getValidEntityName(bucketName);
+            if (!databaseBuckets.containsKey(validName)) {
+                loadedEntitiesList.add(new LoadedEntities(validName));
+                databaseBuckets.put(validName, bucketName);
+            }
+        }
+        return ImmutableList.copyOf(databaseBuckets.keySet());
+    }
```

Comment:
I don't think this is correct.
Not all root buckets are going to be hive stores.
Additionally there could be databases nested under a non root bucket as well, right?

I haven't had time yet to come up with an alternative, but just wanted to call this out first.

Reply:
Another thought I just had is that it seems like you guys should consider maybe integrating with Glue for this.

Since this is exactly the reason why the Hive metadata store exists, so that people can express the location of their data and schema of their data.

Again, this is just me commenting on this from just glancing, so I could be missing something.

If you do go down the route of integrating with Glue, I just wanted to call out that you can coerce the existing data that is read in by the Dataset api into using a new Schema by just using this constructor:

VectorSchemaRoot(Schema schema, List<FieldVector> fieldVectors, int rowCount)

 - Where you pass in the new schema from Glue and then use the existing fieldVectors, and row count, from an existing VectorSchemaRoot.
 -  Also note that we already have a GlueMetadataHandler that you can extend that will already automatically read the Glue schema string and convert it to an Arrow Schema for you, see: `GlueMetadataHandler.java:doGetTable(BlockAllocator blockAllocator, GetTableRequest request, TableFilter filter)` line 422

Reply:
We'll integrate Glue as a future enhancement